Add config options to big scrape, separate scraping and writing within big scrape.

lfashby opened the pull request:
(Sorry for the wall of text...)
These changes are meant to address issues #66, #67, #68  as well as a few suggestions made in the comments of pull request #61.

Here are the larger changes introduced by this pull request:
- Separated the scraping and writing part of `scrape.py` (formerly `scrape_and_write.py`).
  - `write.py` now generates the README table by inspecting the contents of the `tsv/` directory. In the process it also creates a tsv `readme_tsv.tsv` with similar information as is in the README table.

- Added `no_stress`, `no_syllable_boundaries`, and `cut_off_date` options to languages in `languages.json`.
  - Modified `codes.py` to specify default values for these options when adding new languages to `languages.json` and to copy over previously set values for these options.
  - `cut_off_date` should now be set in codes.py prior to running codes.py, I’ve updated the README in languages/wikipron with those instructions.

- Added `dialect` config option (and `require_dialect_label` option) to English, Spanish and Portuguese. 
  - Restructured `scrape.py` to handle when one or more dialects are specified for a language. (Ran this new code on Portuguese, because it is a smaller language, to generate some sample data.)
  - README table now includes dialect information in Wiktionary language name column

The only small changes worth noting are:
- Logging in `scrape.py` will now also output to `scraping.log` which I’ve added to `.gitignore`. This way finding the languages that failed to be scraped is a bit easier (don’t need to scroll through the console). It also outputs the language dict from `languages.json` in the error message for languages that failed to be scraped within our set amount of retries, so it is a bit easier to build a temporary `languages.json` with the failed languages. 
- `scrape.py` will now remove files with less than 100 entries. (TSVs with less than 100 entries have been removed.)

------------------------
I have a few questions regarding dialects that I'd like your thoughts on: 

- How dialects are handled in `languages.json`.
  - I added dialects to `languages.json` in the following way: 
    ```
    "por": {
        ...
        "require_dialect_label": true,
        "dialect": {
            "_bz": "Brazil",
            "_po": "Portugal"
        },
        ...
    },
    ```
  - The keys (`_bz`, `_po`) in `dialect` serve as a sort of extension when naming the dialect tsv files (`por_bz_phonetic.tsv`, for example) and help with easy access to the dialect strings (`"Brazil"`) in `write.py`. If you'd like me to change any of the keys because you'd like different extensions for certain dialects let me know. They can be longer than two letters. I'll provide links to the English, Spanish, and Portuguese entries in `languages.json` as a separate comment so you can review the keys and dialect strings I'm using.
  - Is there a process for finding which dialects are frequently used within a given Wiktionary language category? (Aside from just checking entries and seeing whether dialect information is specified.)

- How dialects are handled in `scrape.py`.
  - As written `scrape.py` will first scrape a language entirely and then scrape for dialects if any are specified. This means it will scrape `por` (Portuguese) with no dialect, then `por` with `"Brazil"` as the dialect and then `por` with `"Portugal"` as the dialect. Is there any reason to scrape for `por` with no dialect when we are specifying a dialect? Do we want to keep the tsvs generated from previously scraping `por` (or `eng`/`spa`) with no dialect?
  - Within `scrape.py`, I moved a lot of what was in `main()` to a separate function in order to handle dialects. There may well be a better way of handling dialects than the way I've tried to do it and I'm open to suggestions on how to improve it.

lfashby commented:
These are what the English, Spanish and Portuguese entries look like in `languages.json`.

https://github.com/kylebgorman/wikipron/blob/faee11782214894fced4e370bc043b8792ac575a/languages/wikipron/src/languages.json#L362-L375

https://github.com/kylebgorman/wikipron/blob/faee11782214894fced4e370bc043b8792ac575a/languages/wikipron/src/languages.json#L1290-L1303

https://github.com/kylebgorman/wikipron/blob/faee11782214894fced4e370bc043b8792ac575a/languages/wikipron/src/languages.json#L1159-L1172

If you'd like, for example, the `"_ca"` in Spanish to be `"_cs"` or any other changes let me know.

lfashby commented:
[This is a link to the README table](https://github.com/kylebgorman/wikipron/blob/faee11782214894fced4e370bc043b8792ac575a/languages/wikipron/tsv/README.md) so you can see how Portuguese looks with dialect info in the Wiktionary name column. I don't think it would make sense to add another column for dialect info as the table is already quite wide and it would only apply to a few languages. 

[This is a link to the tsv of the README table](https://github.com/kylebgorman/wikipron/blob/faee11782214894fced4e370bc043b8792ac575a/languages/wikipron/src/readme_tsv.tsv). It contains mostly the same information as is in the README, but just lists the file name in the link column instead of a relative path in a markdown link. I could potentially add headers if that is useful. A separate column for dialects may be relevant in this format.

kylebgorman commented:
Just wanted to comment on some of the API choices before I dive into code. This is a lot of stuff at once, congrats on getting it all together so fast.

* I don't care for the name `readme_tsv.tsv`; it duplicates "tsv" (duh) and the fact that it is the same data as the README table is irrelevant to me. What is relevant to me as a reader is that it's a TSV file that contains the counts of data for all the languages. (`languages.tsv` maybe?)
* Can you rename `write.py` to something more informative? Maybe `summarize.py` or `generate_summary.py` since that's what it does?
* I personally have no use for the "all-dialectal" versions of English, Portuguese, and Spanish, and if someone wants them, they should just concatenate (and sort) the dialectal ones, right? So I'd suggest that if `dialect` is specified it should pre-empt an all-dialects scrape. (If someone really wanted one they could add THAT to the `dialect` specification in `languages.json`.)
* Your approach to adding dialect suffixes is fine but I would suggest that your code should add the `_` between the language and the dialect code, rather than including it in the JSON file, which feels like an abstractional leak. So it would look like 

     "dialect": { 
         "us": "US | General American", 
         "uk": "UK | Received Pronunciation" 
     }, 

but would still produce `en_us_phonemic.tsv` etc.

kylebgorman commented on languages/wikipron/src/codes.py:
```diff
@@ -104,12 +113,18 @@ def _scrape_wiktionary_info(lang_title):
+        "casefold": None,
+        "no_stress": True,
+        "no_syllable_boundaries": True,
+        "cut_off_date": CUT_OFF_DATE
```
apply reflow-er? I see there's no trailing comma here ;)

kylebgorman commented on languages/wikipron/src/codes.py:
```diff
@@ -104,12 +113,18 @@ def _scrape_wiktionary_info(lang_title):
     with open("iso-639-3_20190408.tsv", "r") as source:
         iso_list = csv.reader(source, delimiter="\t")
-        for (lang_page_title, total_pages) in _cat_members():
+        for (lang_page_title) in _cat_members():
```
remove extraneous parens here (or I think reflow-er will take care of that)

kylebgorman commented on languages/wikipron/src/codes.py:
```diff
@@ -30,7 +31,13 @@
+# Set CUT_OFF_DATE below.
+# CUT_OFF_DATE can be no later than the date
+# at which you plan to begin scraping.
+CUT_OFF_DATE = "2019-11-01"
```
I'm a little confused why this is given as a default here. Why not use `datetime.datetime.today().isoformat()` or something? (For our particular case where we really want to do a cut-off-date'd scrape we can just search-and-replace on` languages.json`, right?)

lfashby replied:
For some reason I interpreted #68 as “Language entries in languages.json should contains all config options that do not use default values set in config.py.” 

So I guess I wasted some time doing that. It would be simpler if the language entries looked like they were before: 
```
    "est": {
        "iso639_name": "Estonian",
        "wiktionary_name": "Estonian",
        "wiktionary_code": "et",
        "casefold": true
    },
```
Or this if a dialect is used:
```
    "eng": {
        "iso639_name": "English",
        "wiktionary_name": "English",
        "wiktionary_code": "en",
        "casefold": true,
        "dialect": {
            "us": "US | General American",
            "uk": "UK | Received Pronunciation"
        }
    },
```

And then `config_settings` in `scrape.py` would handle setting `no_stress`, `no_syllable_boundaries`, and `cut_off_date`, since they are always the same. Then we can just set `cut_off_date` with `datetime` in `scrape.py`.

kylebgorman commented on languages/wikipron/README.md:
```diff
@@ -2,24 +2,26 @@
 ====================
 
```
minor generic comment throughout: `TSV` not `tsv` (at least in comments---snake_case if it's a variable name, sure.)

kylebgorman commented on languages/wikipron/src/codes.py:
```diff
@@ -21,6 +21,7 @@
 * Dialect information may also need to be added manually
 """
 
+# TODO Generate and use a lookup table for the iso639-3 tsv file
```
Generic minor comment throughout this change: `TSV` when you're talking about it in a comment (okay to have a variable name like `tsv` cuz snake_casing conventions).

kylebgorman commented on languages/wikipron/src/scrape.py:
```diff
@@ -66,99 +30,108 @@ def _call_scrape(lang, config, filename_suffix):
+    # Removes TSV files with less than 100 entries.
+    if phonemic_count < 100:
+        logging.info(
+            '"%s", has less than 100 entries in phonemic transcription.',
```
what's with the comma in the log statement? This is going to say:

`Amharic, has less than 100 entries...`

Is that intended?

kylebgorman commented on languages/wikipron/src/scrape.py:
```diff
@@ -66,99 +30,108 @@ def _call_scrape(lang, config, filename_suffix):
     logging.basicConfig(
         format="%(filename)s %(levelname)s: %(asctime)s - %(message)s",
+        handlers=[
+            logging.FileHandler("scraping.log", mode="w"),
```
possibly put filename path here into a module-level constant

kylebgorman commented on languages/wikipron/src/write.py:
```diff
@@ -0,0 +1,97 @@
+
+
+def _readme_insert(wiki_name, row_string):
+    with open("../tsv/README.md", "r+") as source:
```
suggest this path as a module-level constant

kylebgorman commented on languages/wikipron/src/write.py:
```diff
@@ -0,0 +1,97 @@
+            # Rewrite the entire README.
+            source.seek(0)
+            source.truncate()
+            source.write("".join(readme_list))
```
If each element in `readme_list` was a line w/o newlines you could use the more-efficient `writelines` (which takes an iterator of strings and doesn't build up the whole big string at once). But this part is so complex I'm afraid to ask you to kick the hornet's nest...

lfashby replied:
I'll look into this.

kylebgorman commented on languages/wikipron/src/write.py:
```diff
@@ -0,0 +1,97 @@
+                file.rindex("_") + 1 : file.index(".")
+            ].capitalize()
+            wiki_name = languages[iso639_code]["wiktionary_name"]
+            # Assumes we will not remove eng and spa tsv files
```
Is this comment still current? Will it become irrelevant shortly? If so remove.

kylebgorman commented on languages/wikipron/src/write.py:
```diff
@@ -0,0 +1,97 @@
+        def sorting(ele):
+            return ele[3]
+        readme_tsv_list.sort(key=sorting)
+        for lang_row in readme_tsv_list:
```
Strongly suggest you use `csv.writer` for this (you just need to tell it you're using a "\t" separator). Then call it's `writerows` method with `readme_tsv_list` as an argument and you're free of this loop too!

lfashby commented:
Okay, I'll start working on this tomorrow and just reply to a few comments now.

lfashby commented:
In addition to changing the name of `readme_tsv.tsv` I think I should also change where it gets outputted. Right now it just gets placed in the src directory, but I presume the tsv directory is the more appropriate place.

kylebgorman commented:
Not sure. Maybe the parent directory of src/ and tsv/ ? That way there’s no
risk of it being confused with an actual language file.

K

On Thu, Oct 24, 2019 at 11:12 PM Lucas Ashby <notifications@github.com>
wrote:

> In addition to changing the name of readme_tsv.tsv I think I should also
> change where it gets outputted. Right now it just gets placed in the src
> directory, but I presume the tsv directory is the more appropriate place.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kylebgorman/wikipron/pull/76?email_source=notifications&email_token=AABG4OODNAGTVWX5L3BSDFDQQJP2DA5CNFSM4JE44HH2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECHBMEQ#issuecomment-546182674>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABG4OMJZSK6JAITHWZN2J3QQJP2DANCNFSM4JE44HHQ>
> .
>

lfashby commented:
I think these changes touch on all the comments Kyle made. Aside from formatting and file name stuff, the big differences are: 

- I renamed and reworked `write.py` (now `generate_summary.py`) to be more efficient. Formally it would rewrite the README (by reading it in, splitting the rows into a list of lists, inserting a new row, converting the list of lists into a big string and writing that string) every time we iterated through a file in the tsv directory. Now it just constructs a list of lists once, converts those internal lists to strings once, and writes an all new README once. It's perhaps still not ideal, but I feel better about it.

- I added what I thought were important paths as constants to `codes.py`.

- I removed English, Spanish, and Portuguese TSVs that don't use a dialect (in preparation for the next scrape).

kylebgorman commented on languages/wikipron/src/generate_summary.py:
```diff
@@ -0,0 +1,76 @@
+
+    # Sort by wiktionary language name,
+    # with phonemic entries before phonetic
+    def sorting(ele):
```
still just use a lambda here.

lfashby replied:
As in: 
```
languages_summary_list.sort(key=lambda ele: ele[3] + ele[5])
readme_list.sort(key=lambda ele: ele[3] + ele[5])
```
Or something along the lines of: 
```
f = lambda ele: ele[3] + ele[5]
languages_summary_list.sort(key=f)
readme_list.sort(key=f)
```
The later is 'DRY-er' but isn't assigning a lambda to a variable an anti-pattern?

kylebgorman commented on languages/wikipron/src/scrape.py:
```diff
@@ -64,101 +29,109 @@ def _call_scrape(lang, config, filename_suffix):
-            )
+        if os.path.exists(phonemic_path):
             os.remove(phonemic_path)
+        if os.path.exists(phonetic_path):
```
Just noticed this. This is a "dark pattern" because it is theoretically possible for the file to blink in or out of existence between the check for existence and the remove. The recommended pattern is instead to do a `try: os.remove(...)` and then catch and ignore the exception (`OSError` probably) if it fails. I realize this is an annoying one...

lfashby replied:
Found this from googling:
```
from contextlib import suppress

with suppress(OSError):
    os.remove(filename)
```
Seems nicer than ignoring the exception. Let me know if you have a preference. Is it the `os.path.exists()` check that makes the file potentially blink in or out of existence? Do I need to add this to all `os.remove()` statements or just the ones that involve `os.path.exists()`

kylebgorman commented on languages/wikipron/src/scrape.py:
```diff
@@ -64,101 +29,109 @@ def _call_scrape(lang, config, filename_suffix):
-            phonemic_readme_row_string = (
-                "| " + " | ".join(phonemic_row) + " |\n"
+        return
+    # Remove files for languages that returned nothing
```
period for consistency on this comment I suppose

kylebgorman commented on languages/wikipron/src/scrape.py:
```diff
@@ -64,101 +29,109 @@ def _call_scrape(lang, config, filename_suffix):
-                "| " + " | ".join(phonemic_row) + " |\n"
+        return
+    # Remove files for languages that returned nothing
+    elif phonemic_count == 0 and phonetic_count == 0:
```
probably `if not phonemic_count and not phonetic_count` or something like that.

kylebgorman commented:
Former is better and hopefully the reflower can make it pretty too.

On Fri, Oct 25, 2019 at 6:06 PM Lucas Ashby <notifications@github.com>
wrote:

> *@lfashby* commented on this pull request.
> ------------------------------
>
> In languages/wikipron/src/generate_summary.py
> <https://github.com/kylebgorman/wikipron/pull/76#discussion_r339255133>:
>
> > +
> +        row = [
> +            iso639_code,
> +            languages[iso639_code]["iso639_name"],
> +            wiki_name,
> +            str(languages[iso639_code]["casefold"]),
> +            transcription_level,
> +            str(num_of_entries),
> +        ]
> +        # TSV and README have different first column.
> +        languages_summary_list.append([file] + row)
> +        readme_list.append([f"[TSV]({file})"] + row)
> +
> +    # Sort by wiktionary language name,
> +    # with phonemic entries before phonetic
> +    def sorting(ele):
>
> As in:
>
> languages_summary_list.sort(key=lambda ele: ele[3] + ele[5])
> readme_list.sort(key=lambda ele: ele[3] + ele[5])
>
> Or something along the lines of:
>
> f = lambda ele: ele[3] + ele[5]
> languages_summary_list.sort(key=f)
> readme_list.sort(key=f)
>
> The later is 'DRY-er' but isn't assigning a lambda to a variable an
> anti-pattern?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kylebgorman/wikipron/pull/76?email_source=notifications&email_token=AABG4OO4CCTKSXRH4HM4I5LQQNUWHA5CNFSM4JE44HH2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCJJ2WVI#discussion_r339255133>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABG4OP2FUYEI4PUWANXOZDQQNUWHANCNFSM4JE44HHQ>
> .
>

lfashby commented:
I'm not sure what reflower or reflow-er is and searching 'reflower python' doesn't seem to bring up anything useful.

kylebgorman commented:
Sorry, I’m referring to ‘white’.

On Fri, Oct 25, 2019 at 6:24 PM Lucas Ashby <notifications@github.com>
wrote:

> I'm not sure what reflower or reflow-er is and searching 'reflower python'
> doesn't seem to bring up anything useful.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kylebgorman/wikipron/pull/76?email_source=notifications&email_token=AABG4OMEHVGLQ73H36WCWBLQQNW3DA5CNFSM4JE44HH2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECJWZMQ#issuecomment-546532530>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABG4OJLQ6YLZ4RRIR6JZZ3QQNW3DANCNFSM4JE44HHQ>
> .
>

lfashby commented:
I made the above changes and ran white. There were a few things I could do with the `os.remove` try-except block but I just went with the most straightforward method and added some comments above it.

jacksonllee commented on languages/wikipron/README.md:
```diff
@@ -2,17 +2,18 @@
-Wiktionary languages with over 100 entries. It writes the results of those
-scrape calls to TSVs and generates a [README](./tsv/README.md) with selected
-information regarding the contents of those TSVs and the configuration settings
+Wiktionary languages with over 100 entries. [write.py](./src/write.py)
```
`write.py` -> `generate_summary.py`?

jacksonllee commented on languages/wikipron/README.md:
```diff
@@ -30,9 +31,8 @@ Steps used to update the dataset
+        or search for `null` values within 
+        [languages.json](./src/languages.json).
 2.  Run [scrape.py](./src/scrape.py).
+3.  Run [write.py](./src/write.py).
```
`generate_summary.py`

jacksonllee commented on languages/wikipron/src/generate_summary.py:
```diff
@@ -0,0 +1,83 @@
+    readme_list = []
+    languages_summary_list = []
+    path = "../tsv"
+    for file in os.listdir(path):
```
`file` -> `file_path` or something?

jacksonllee commented on languages/wikipron/tsv/por_bz_phonetic.tsv:
```diff
@@ -0,0 +1,366 @@
+adaga	a d a ɡ a
+adenina	a d e n i n a
+adeus	a d e w s
+adição	a d͡ ʒ i s ɐ̃ w
```
Oops, we forgot about having to correctly handle the tie bar for affricates. I'm opening an issue for this now.

lfashby commented:
> boolean coercion here?

You mean like `if not phonemic_count and not phonetic count` as in the if block below it?

I could potentially set up` _call_scrape()` so it always returns an int and then remove the files when that int is less than 100, but the idea behind having these separate if blocks and the reason `_call_scrape()` can return either an int or None is so we can log the different ways languages can fail.

kylebgorman commented:
I don't know if this is a good case for a returning heterogeneous types. If
you want to do that any of the following would be preferable IMO:

1. raise an exception
2. return 0 as a default
3. return -1 if it's important to distinguish between that and 0

I think #2 is probably the best option.


On Sun, Oct 27, 2019 at 11:21 AM Lucas Ashby <notifications@github.com>
wrote:

> boolean coercion here?
>
> You mean like if not phonemic_count and not phonetic count as in the if
> block below it?
>
> I could potentially set up _call_scrape() so it always returns an int and
> then remove the files when that int is less than 100, but the idea behind
> having these separate if blocks and the reason _call_scrape() can return
> either an int or None is so we can log the different ways languages can
> fail.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kylebgorman/wikipron/pull/76?email_source=notifications&email_token=AABG4OPT2HCZSGQSSDUG3ALQQWWWFA5CNFSM4JE44HH2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECLAXQA#issuecomment-546704320>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABG4OKCAGFYCPB5RVHTR6TQQWWWFANCNFSM4JE44HHQ>
> .
>

kylebgorman commented on languages/wikipron/README.md:
```diff
@@ -2,17 +2,18 @@
 [languages.json](./src/languages.json). It queries Wiktionary for all languages
 with over 100 entries. It also outputs
-[failed\_langauges.json](./src/failed_languages.json), a list of languages on
+[unmatched\_langauges.json](./src/unmatched_languages.json), a list of languages on
```
`langauges`

kylebgorman commented on languages/wikipron/src/codes.py:
```diff
@@ -104,12 +113,12 @@ def _scrape_wiktionary_info(lang_title):
+    unmatched_languages = {}
+    with open(LANGUAGES_PATH, "r") as source:
         prev_languages = json.load(source)
     with open("iso-639-3_20190408.tsv", "r") as source:
```
Model level constant?

kylebgorman commented on languages/wikipron/src/codes.py:
```diff
@@ -122,40 +131,53 @@ def main():
-        failed_dict = json.dumps(failed_languages, indent=4)
-        failed.write(failed_dict)
+    # All languages that failed to be matched with data in ISO 639 TSV file.
+    with open("unmatched_languages.json", "w") as unmatched:
```
module-level constant

kylebgorman commented on languages/wikipron/src/generate_summary.py:
```diff
@@ -0,0 +1,85 @@
+        languages_summary_list.append([file_path] + row)
+        readme_list.append([f"[TSV]({file_path})"] + row)
+
+    # Sort by wiktionary language name,
```
`Wiktionary`

kylebgorman commented on languages/wikipron/src/generate_summary.py:
```diff
@@ -0,0 +1,85 @@
+
+    # Sort by wiktionary language name,
+    # with phonemic entries before phonetic.
+    languages_summary_list.sort(key=lambda ele: ele[3] + ele[5])
```
Okay, so now you're using this twice, so probably define it outside of main as `_3_and_5` or something like that?

kylebgorman commented on languages/wikipron/src/scrape.py:
```diff
@@ -64,101 +29,92 @@ def _call_scrape(lang, config, filename_suffix):
+    return 0
+
+
+def build_config_and_filter_scrape_results(
```
maybe too long, also maybe you want `_` at the beginning

kylebgorman commented:
Other than that, looks good to me.
